File: nodejs/selenium-proxy-location/scrape.js
#!/usr/bin/env node
const { Builder, Browser, By } = require('selenium-webdriver');
const {
    AUTH = 'USER:PASS',
    TARGET_URL = 'https://geo.brdtest.com/mygeo.json',
    LOCATION = 'amsterdam',
} = process.env;

const LOCATIONS = Object.freeze({
    amsterdam: { lat: 52.377956, lon: 4.897070 },
    london: { lat: 51.509865, lon: -0.118092 },
    new_york: { lat: 40.730610, lon: -73.935242 },
    paris: { lat: 48.864716, lon: 2.349014 },
});

async function scrape(url = TARGET_URL, location = LOCATION) {
    if (AUTH == 'USER:PASS') {
        throw new Error(`Provide Scraping Browsers credentials in AUTH`
            + ` environment variable or update the script.`);
    }
    if (!LOCATIONS[location]) {
        throw new Error(`Unknown location`);
    }
    const { lat, lon } = LOCATIONS[location];
    console.log(`Connecting to Browser...`);
    const server = `https://${AUTH}@brd.superproxy.io:9515`;
    const driver = await new Builder()
        .forBrowser(Browser.CHROME)
        .usingServer(server)
        .build();
    try {
        console.log(`Connected! Changing proxy location`
            + ` to ${location} (${lat}, ${lon})...`);
        await driver.sendAndGetDevToolsCommand('Proxy.setLocation', {
            lat, lon,
            distance: 50 /* kilometers */,
        });
        console.log(`Navigating to ${url}...`);
        await driver.get(url);
        console.log(`Navigated! Scraping data...`);
        const body = await driver.findElement(By.css('body'));
        const data = await body.getText();
        console.log(`Scraped! Data:`, JSON.parse(data));
    } finally {
        await driver.quit();
    }
}

if (require.main == module) {
    scrape().catch(error => {
        console.error(error.stack || error.message || error);
        process.exit(1);
    });
}

File: python/selenium-proxy-location/scrape.py
#!/usr/bin/env python3
import json
from os import environ
from selenium.webdriver import Remote, ChromeOptions as Options
from selenium.webdriver.chromium.remote_connection import ChromiumRemoteConnection as Connection
from selenium.webdriver.common.by import By

AUTH = environ.get('AUTH', default='USER:PASS')
TARGET_URL = environ.get('TARGET_URL',
                         default='https://geo.brdtest.com/mygeo.json')
LOCATION = environ.get('LOCATION', default='amsterdam')
LOCATIONS = {
    'amsterdam': (52.377956, 4.897070),
    'london': (51.509865, -0.118092),
    'new_york': (40.730610, -73.935242),
    'paris': (48.864716, 2.349014),
}


def scrape(url=TARGET_URL, location=LOCATION):
    if AUTH == 'USER:PASS':
        raise Exception('Provide Scraping Browsers credentials in AUTH ' +
                        'environment variable or update the script.')
    if location not in LOCATIONS:
        raise Exception('Unknown location')
    lat, lon = LOCATIONS[location]
    print('Connecting to Browser...')
    server_addr = f'https://{AUTH}@brd.superproxy.io:9515'
    connection = Connection(server_addr, 'goog', 'chrome')
    driver = Remote(connection, options=Options())
    try:
        print('Connected! Changing proxy location'
              + f'to {location} ({lat}, {lon})...')
        driver.execute('executeCdpCommand', {
            'cmd': 'Proxy.setLocation',
            'params': {
                'lat': lat, 'lon': lon,
                'distance': 50  # kilometers
            },
        })
        print(f'Navigating to {url}...')
        driver.get(url)
        print('Navigated! Scraping data...')
        body = driver.find_element(By.CSS_SELECTOR, 'body')
        data = body.text
        print(f'Scraped! Data: {json.loads(data)}')
    finally:
        driver.quit()


if __name__ == '__main__':
    scrape()
